### ISSUE TITLE
"Did not find end of central directory signature"

### ISSUE THREAD

**Issue by trhowitaway** [jgm/zip-archive#25]
Hello, I am having trouble unzipping this archive http://rghost.net/8pkKrx4sY , it fails with `Did not find end of central directory signature`, while `unzip` decompress it just fine.

This is the output of zipinfo:

```
u@m:~/folder$ zipinfo nonsense.zip
Archive:  nonsense.zip
Zip file size: 239899 bytes, number of entries: 3
-rw-rw-rw-  2.0 fat    82798 b- defX 00-Nov-25 20:34 NONSENSE.MZX
-rw-rw-rw-  2.0 fat   110702 b- defN 80-Jan-01 00:00 tomsdine.mod
-rw-rw-rw-  2.0 fat   200302 b- defN 96-Oct-08 10:59 CD_STAR.MOD
3 files, 393802 bytes uncompressed, 239574 bytes compressed:  39.2%
```

So `defX` and `defN`, which are deflate methods (normal & maximum compression).

Documentation states:

> [t]here is no support for [...] compression methods other than Deflate.

So I decided to open a bug report.

**Comment by jgm** [comment]
Since this is a documented limitation, this isn't really a bug.

I'll change the title to a request to support other compression methods.  Deflate is the only one I could see how to support at the time I wrote this library.

I'm not inclined to work on this, since zip-archive works well enough for my purposes, but if someone else does, I'd consider a PR.

**Comment by jgm** [comment]
I see -- I initially thought these were deflation methods other than Deflate, but if they are versions of deflate, I'll have to look into this.

**Comment by jgm** [comment]
It looks as if
https://hackage.haskell.org/package/zlib-0.6.1.1/docs/Codec-Compression-Zlib.html
allows you to specify `compressLevel`.
Would it be right to map `defX` to `NormalCompression` and `defN` to `BestCompression`?

**Comment by jgm** [comment]
@trhowitaway I can no longer retrieve the zip archive from the link you provided.
Can you make it available again?

**Comment by stephen-smith** [comment]
Was this fixed?  I'm getting something similar with some zip files I can't share, and I was hoping an upgrade to zip-archive would fix it.

If this wasn't fixed, I'll try to generate a example file that I can share.

**Comment by jgm** [comment]
It doesn't seem to have been fixed.
I believe it was closed because I didn't have a way to reproduce the problem.
If you can provide an example, that would be helpful!

**Comment by stephen-smith** [comment]
> Would it be right to map `defX` to `NormalCompression` and `defN` to `BestCompression`?

I think defN is Normal and defX is Best and defF is Fast.  With zip on my Debian system, `-0` is "stor", `-1` and `-2` are "defF", `-3` through `-7` are "defN", and `-8` and `-9` are "defX".

**Comment by stephen-smith** [comment]
Is it possible that `zip warning: undefined bits used in flags = 0x0808` that I get from `zip -v -T` could be causes zip-archive to fail?  I'm not (yet) able to have `zip` output something that zip-archive doesn't like, but the Java process that's creating the file I need to consume might be doing something silly.  I'll have to dig through that zip implementation (or at least the driver code) to see.

**Comment by stephen-smith** [comment]
Here's an example file that zip-archive gives me the "Did not find end of central directory signature" error on.  It's the smallest one I've gotten to fail.

[zip-archive_bad.zip](https://github.com/jgm/zip-archive/files/3758146/zip-archive_bad.zip)

```
% zipinfo zip-archive_bad.zip
Archive:  zip-archive_bad.zip
Zip file size: 9629882 bytes, number of entries: 2
-rw----     2.0 fat  9597229 bl defN 19-Oct-22 14:43 SMNT.ZIP
-rw----     2.0 fat    33570 bl defN 19-Oct-22 14:43 UMNT.ZIP
2 files, 9630799 bytes uncompressed, 9629644 bytes compressed:  0.0%
% zip -T zip-archive_bad.zip
test of zip-archive_bad.zip OK
```

All the data in the files is just pulled from /dev/urandom.  All the file names have been anonymized.  Otherwise, it's pretty close to my use case.  If I fill all the files from /dev/zero instead /dev/urandom, zip-archive is able to process the result, not sure if that's relevant.

I'm using an old version of zip-archive (0.2.3.4), as I'm building on a Debian Jessie system in order to match the glibc version of the embedded system I'm wanting to deploy to.

**Comment by stephen-smith** [comment]
[Smaller example](https://github.com/jgm/zip-archive/files/3758836/zip-archive_bad_small.zip)

**Comment by stephen-smith** [comment]
[Smaller Example](https://github.com/jgm/zip-archive/files/3758859/zip-archive_bad_small.zip)

**Comment by jgm** [comment]
I can reproduce with the latest dev version.

**Comment by hanjoosten** [comment]
I am running into this problem too. I use zip-archive 0.4.1. Any chance that this could be looked into?

**Comment by hanjoosten** [comment]
```
2020-03-04 21:32:08.391517: [debug] Start extraction of prototype framework.
Error encountered during deployment of prototype framework:
  Failed to extract the archive found at https://github.com/AmpersandTarski/Prototype/archive/v1.3.1\.zip
Data.Binary.Get.runGet at position 4: Did not find end of central directory signature
CallStack (from HasCallStack):
  error, called at libraries\\binary\\src\\Data\\Binary\\Get.hs:351:5 in binary-0.8.6.0:Data.Binary.Get
Error encountered during deployment of prototype framework

**Comment by jgm** [comment]
Looks like with the smaller example from @stephen-smith , it is failing because the test at line 790 fails:  the compressed size read from the archive doesn't match the length of the compressed data.  Need to investigate further.

**Comment by hanjoosten** [comment]
tldr: The problem was caused by an error on my side, not by the library. Sorry.

longer story:
I have been doing some diagnosis, to see if there were any differences in the versions of libraries between two branches of my project. Our development branch has no problem with the archive, and an feature branch gives the error I mentioned two days ago. 

The feature branch is about upgrading to pandoc 2.9 including the new types. This causes loads of String stuff to be converted to Text. In doing so, I accidentally messed up the URL of the file to download. The error message above shows the URL. Still strange that the download succeeded but the archive couldn't be extracted.

If the same file is downloaded with the url 
```
https://github.com/AmpersandTarski/Prototype/archive/v1.3.1.zip
```
there is no problem.


Sorry for wasting your time.

**Comment by stephen-smith** [comment]
Can we reopen?  The new activity from @hanjoosten might not need a fix, but my examples from 2019-10-22 are both zip files that I expect would extract cleanly (and they do in other tools), but "latest dev" as of 2019-10-23 fails on them.

**Comment by jgm** [comment]
Sorry, I wasn't tracking that there were two separate examples here. I was actually only looking at @stephen-smith's anyway.  That's the one that fails because the reported compressed length is different from the actual length of the data.  I don't know what could cause that, so I'm at a loss.

**Comment by jgm** [comment]
See Test-KO.docx at <https://groups.google.com/d/msgid/pandoc-discuss/CAO18McU3P3q0N0cVRXoq-nh1ZLZhuTM7uWdWkEVRqe22y4%3DKAQ%40mail.gmail.com> for another example that fails with a similar error.

```
% zip-archive -d ../pandoc/Test-KO.docx
zip-archive: Data.Binary.Get.runGet at position 6819: getWordsTilSig: signature not found before EOF
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:351:5 in binary-0.8.6.0:Data.Binary.Get
```

**Comment by pkamenarsky** [comment]
It seems that the [zip](https://hackage.haskell.org/package/zip) package can decrypt @stephen-smith's example, maybe this will help.

**Comment by memeplex** [comment]
I'm also getting this from a docx inside a zip that Google Drive produced while exporting from its native format.

**Comment by memeplex** [comment]
Office365 is unable to open it, anyway, so I guess Google is to blame here. If I instead download a single file (not a directory, which zips all files inside) the produced docx is smaller and works just fine.

**Comment by jgm** [comment]
It looks like this was fixed by #66.
I can now decode both of the sample zips above.
